docs/pallas/quickstart.md

```diff
@@ -0,0 +1,230 @@
+import jax
+from jax.experimental import pallas as pl
+import jax.numpy as jnp
+import numpy as np
+```
+
+We'll first write the "hello world" in Pallas, a kernel that adds two vectors.
+
+```{code-cell} ipython3
+def add_vectors_kernel(x_ref, y_ref, o_ref):
+  x, y = x_ref[...], y_ref[...]
+  o_ref[...] = x + y
+```
+
+**`Ref` types**
+
+Let's dissect this function a bit. Unlike most JAX functions you've probably written, it does not take in `jax.Array`s as inputs and doesn't have any outputs. Instead it takes in *`Ref`* objects as inputs. Note that we also don't have any outputs but we are given an `o_ref`, which corresponds to the desired output.
+
+**Reading from `Ref`s**
+
+In the body, we are first reading from `x_ref` and `y_ref`, indicated by the `[...]` (the ellipsis means we are reading the whole `Ref`; alternatively we also could have used `x_ref[:]`). Reading from a `Ref` like this returns a `jax.Array`.
+
+**Writing to `Ref`s**
+
+We then write `x + y` to `o_ref`. Mutation has not historically been supported in JAX -- `jax.Array`s are immutable! `Ref`s are new (experimental) types that allow mutation under certain circumstances. We can interpret writing to a `Ref` as mutating its underlying buffer.
+
++++
+
+So we've written what we call a "kernel". It's a program that will be executed on an accelerator. How do we invoke it from a JAX computation? We use the `pallas_call` higher-order function.
+
+```{code-cell} ipython3
+@jax.jit
+def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
+  return pl.pallas_call(add_vectors_kernel,
+                        out_shape=jax.ShapeDtypeStruct(x.shape, x.dtype)
+                        )(x, y)
+add_vectors(jnp.arange(8), jnp.arange(8))
+```
+
+`pallas_call` takes in a kernel function as input and some other metadata about it. Here we specify `out_shape`, an object that has a `.shape` and `.dtype` (or a list thereof).
+`out_shape` determines the shape/dtype of `o_ref` in our `add_vector_kernel`.
+
+`pallas_call` returns a function that takes in and returns `jax.Array`s.
+
++++
+
+**What's actually happening here?**
+
+Thus far we've described how to think about Pallas kernels but what we've actually accomplished is we're writing a function that's executed very close to the compute units.
+
+On GPU, `x_ref` corresponds to a value in HBM and when we do `x_ref[...]` we are copying the value from HBM into SRAM (this is a costly operation generally speaking!). We then use GPU vector compute to execute the addition, then copy the resulting value in SRAM back to HBM.
+
+On TPU, we do something slightly different. Before the kernel is ever executed, we fetch the value from HBM into SRAM. `x_ref` therefore corresponds to a value in SRAM and when we do `x_ref[...]` we are copying the value from SRAM into a register. We then use TPU vector compute to execute the addition, then copy the resulting value back into SRAM. After the kernel is executed, the SRAM value is copied back into HBM.
```

Comment:
I think those two paragraphs might belong in backend specific guides.

Reply:
I think a shallow overview seems okay here to at least give the user some mental model of what's happening. We should go into a lot more depth in the platform specific guides.